driver-core/src/main/com/mongodb/internal/time/Timeout.java

```diff
@@ -15,245 +15,127 @@
-     *
-     * @throws AssertionError If the timeout is {@linkplain #isInfinite() infinite}.
-     * @see #remainingOrInfinite(TimeUnit)
+     * @see TimePoint#remaining
      */
-    public long remaining(final TimeUnit unit) {
-        assertFalse(isInfinite());
-        return isImmediate() ? 0 : convertRoundUp(remainingNanos(TimePoint.now()), unit);
-    }
+    long remaining(TimeUnit unit);
 
     /**
-     * Returns a negative value for {@linkplain #isInfinite() infinite} timeouts,
-     * otherwise behaves identically to {@link #remaining(TimeUnit)}.
-     * Use {@link #expired(long)} to check if the returned value signifies that a timeout is expired.
-     *
-     * @see #remaining(TimeUnit)
+     * A convenience method that handles converting to -1, for integration with older internal APIs.
+     * <p>
+     * Returns an {@linkplain #isInfinite() infinite} timeout if {@code durationNanos} is either negative
+     * or is equal to {@link Long#MAX_VALUE}, otherwise a timeout of {@code durationNanos}.
+     * <p>
+     * Note that the contract of this method is also used in some places to specify the behavior of methods that accept
+     * {@code (long timeout, TimeUnit unit)}, e.g., {@link com.mongodb.internal.connection.ConcurrentPool#get(long, TimeUnit)},
+     * so it cannot be changed without updating those methods.</p>
      */
-    public long remainingOrInfinite(final TimeUnit unit) {
+    default long remainingOrNegativeForInfinite(final TimeUnit unit) {
         return isInfinite() ? -1 : remaining(unit);
     }
 
     /**
-     * @see #expired(long)
-     */
-    public boolean expired() {
-        return expired(remainingOrInfinite(NANOSECONDS));
-    }
-
-    /**
-     * Returns {@code true} if and only if the {@code remaining} time is 0 (the time unit is irrelevant).
-     *
-     * @see #remaining(TimeUnit)
-     * @see #remainingOrInfinite(TimeUnit)
-     * @see #expired()
-     */
-    public static boolean expired(final long remaining) {
-        return remaining == 0;
-    }
-
-    /**
-     * @return {@code true} if and only if the timeout duration is considered to be infinite.
+     * @return an infinite (non-expiring) timeout
      */
-    public boolean isInfinite() {
-        return equals(INFINITE);
+    static Timeout infinite() {
+        return TimePoint.infinite();
     }
 
     /**
-     * @return {@code true} if and only if the timeout duration is 0.
+     * @param duration the positive duration, in the specified time unit.
+     * @param unit the time unit.
+     * @return a timeout that expires in the specified duration after now.
      */
-    public boolean isImmediate() {
-        return equals(IMMEDIATE);
-    }
-
-    @Override
-    public boolean equals(final Object o) {
-        if (this == o) {
-            return true;
+    static Timeout expiresIn(final long duration, final TimeUnit unit) {
+        // TODO (CSOT) confirm that all usages in final PR are non-negative
```

Comment:
Note to address all TODOs prior to merging. (All files in github can be loaded using `document.querySelectorAll('.load-diff-button').forEach(node => node.click())` in console, and then ctrl-f for `TODO`)

Reply:
I have addressed critical `//TODO` comments.